internal/mesh/internal/controllers/explicitdestinations/controller.go

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package explicitdestinations
+
+import (
+	"context"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/hashicorp/consul/internal/controller"
+	"github.com/hashicorp/consul/internal/mesh/internal/mappers/workloadselectionmapper"
+	"github.com/hashicorp/consul/internal/mesh/internal/types"
+	"github.com/hashicorp/consul/internal/resource"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v2beta1"
+	pbmesh "github.com/hashicorp/consul/proto-public/pbmesh/v2beta1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+const ControllerName = "consul.io/explicit-destinations-controller"
+
+func Controller(destinationsMapper *workloadselectionmapper.Mapper[*pbmesh.Destinations]) controller.Controller {
+	if destinationsMapper == nil {
+		panic("destinations mapper is required")
+	}
+
+	return controller.ForType(pbmesh.ComputedExplicitDestinationsType).
+		WithWatch(pbmesh.DestinationsType, destinationsMapper.MapToComputedType).
+		WithWatch(pbcatalog.WorkloadType, controller.ReplaceType(pbmesh.ComputedExplicitDestinationsType)).
+		WithReconciler(&reconciler{destinations: destinationsMapper})
+}
+
+type reconciler struct {
+	destinations *workloadselectionmapper.Mapper[*pbmesh.Destinations]
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, rt controller.Runtime, req controller.Request) error {
+	rt.Logger = rt.Logger.With("controller", ControllerName, "id", req.ID)
+
+	// Look up the associated workload.
+	workloadID := resource.ReplaceType(pbcatalog.WorkloadType, req.ID)
+	workload, err := resource.GetDecodedResource[*pbcatalog.Workload](ctx, rt.Client, workloadID)
+	if err != nil {
+		rt.Logger.Error("error fetching workload", "error", err)
+		return err
+	}
+
+	// If workload is not found, the decoded resource will be nil.
+	if workload == nil || workload.GetResource() == nil || workload.GetData() == nil {
+		// When workload is not there, we don't need to manually delete the resource
+		// because it is owned by the workload. In this case, we skip reconcile
+		// because there's nothing for us to do.
+		rt.Logger.Trace("the corresponding workload does not exist", "id", workloadID)
+		return nil
+	}
+
+	// Get existing ComputedDestinations resource (if any).
+	cpc, err := resource.GetDecodedResource[*pbmesh.ComputedExplicitDestinations](ctx, rt.Client, req.ID)
```

Comment:
Nit: What does `cpc` stand for? Maybe `cd` or `ced` would be clearer since they are the initials for the full resource type.

Reply:
great catch! this is a copy-pasta error.